### ISSUE TITLE
Enhancement proposed: to display Open Type layout features.

### ISSUE THREAD

**Issue by roger64** [koreader/koreader#5821]
* KOReader version: all
It's about making use of the _font-variant_ property (or font-variant-_something_ property) to display some or all alternate properties of Open-Type fonts. 

This is a feature Calibre (editor and viewer), and Sigil already display. It's accepted by ePubcheck. It lacks support among commercial ebook-viewers.

Rationale

Implement the use of some OT features like _oldnum_ or _hlig_ which have  a gracious fallback . Users with an advanced viewer like Koreader could use it, and nothing would change for other users (like it happens today with in-page note display).

Here is an example with a joint ePub3 book which makes use of this "oldnum" OT property. You can see it displayed properly with Calibre viewer and/or Sigil. The OT font of this ePub has been subsetted with the relevant Calibre editor tool and the SubsetFont plugin of Sigil.

The font used is a derivative of Linux Libertine.

[villars.zip](https://github.com/koreader/koreader/files/4165093/villars.zip)

**Comment by poire-z** [comment]
Relevant formal syntax: https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant
So, no, thanks :)

I've passed by this when looking at supporting the most used one: smallcaps, but it would be really complicated the way we instantiate/select fonts (which would need a new instance for each variation in harfbuzz features).
So, for me, even if it would be nice, it's really not worth the huge effort.

Other visited links:
https://github.com/ArtifexSoftware/mupdf/commit/0c58f612d2 smallcaps in MuPDF via harfbuzz
https://www.mail-archive.com/harfbuzz@lists.freedesktop.org/msg04588.html
`https://github.com/mapbox/mapbox-gl-js/issues/5547`

**Comment by roger64** [comment]
Thanks for your answer.

I am just a plain user and I have no idea of how difficult a desirable thing like this one may be to implement, so forgive me for this. I have written that Calibre (both editor and viewer) and Sigil can display OT features using font-variant and/or font-variant-_something_  properties. I should have added that they had no work to do since this capability was built-in with their Qt framework.

I understand that it's different for Koreader which makes use of another rendering tool (harfbuzz). 

If this can be useful, I learnt that Prince PDF faced a similar problem some few years ago. Currently, it displays nicely OT features. To achieve this, it made use of "fontkit", following a 2017 proposal (found on a Prince pdf forum)of a knowledgeable user that I quote further:

> it may be of interest to you that fontkit, a modern an extensive OpenType font parser and layout engine work in JS has just received a comprehensive upgrade with a full Indic shaper that is a port of the spec-compliant Harfbuzz shaper:
https://github.com/devongovett/fontkit/pull/99

Perhaps Prince could make use of fontkit (if it's more convenient than using Harfbuzz) via JS so that you could do full OpenType Layout rendering for all Indic scripts. Note that fontkit also supports AAT features and variable OpenType fonts. 

I think I have read that fontkit is a kind of fork of harfbuzz.

Best regards and thanks for your amazing and extensive work!

Roger

**Comment by poire-z** [comment]
Cross ref: https://www.mobileread.com/forums/showthread.php?t=327911

As @NiLuJe answered there, Fontkit is a javascript implementation similar to Harbuzz. We already have Harfbuzz, and we don't have any javascript engine, and don't need any. So, it will not help us.

I'm not trashing the idea, I'll give it thoughts from time to time, until some light bulb moment shows me a clean & simple way to do it :)

Anyway, one thing we'll need is a mapping of font-variant named values to OT features tags (the 4 letter tags).
https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant
https://drafts.csswg.org/css-fonts-3/#propdef-font-variant
    \\/
https://en.wikipedia.org/wiki/List_of_typographic_features
https://docs.microsoft.com/en-us/typography/opentype/spec/features_ko#tag-onum

I haven't found any such mapping on the web. May be a bit of it in that Fontkit file: https://github.com/foliojs/fontkit/blob/master/src/aat/AATFeatureMap.js - but I don't know the diff between AAT and OTF... and these CamelCase values do not look like font-variant hyphen-values.

I'm not interested in learning about OT features - so, if anyone wants to tackle building such a map, feel free to work on it.
We don't need to support all of them, so we'll have to decide which are nice to support and which aren't really needed. And which should be publishers' choices - and which are engine typesetting choices (like +kern and +liga).

For info, we enable or disable some of these tags depending on Font kerning good|best:
https://github.com/koreader/crengine/blob/1a3d236a1454212a8034829dc9e32321253f86d5/crengine/src/lvfntman.cpp#L1104-L1144

As for the implementation in crengine, and I don't see how any external library will help us with that.
We'll need to:
- parse CSS font-variant named values, possibly mapping each supported name to a bit in a 32 or 64 bits bitmap - handle inherited combinations (upper node font-variant > lower node different font-variant - might not be easy, no such thing yet).
- work out how to support this in lvfntman.cpp - different font instances, glyphs caches - and this is where it will be interesting (for me) and where I have no real clue currently.

Anyway, not actively working on it, giving it some thoughts - so may be one day, no promise.

**Comment by shermp** [comment]
Hi @poire-z I've been thinking about the topic this evening actually.

First, AAT stands for Apple Advanced Typography, so probably not relevant for CRengine.

A known list of feature tags can be found [here](https://docs.microsoft.com/en-us/typography/opentype/spec/featurelist)

For example, `tnum` = tabular nums, `dlig` = Discretionary Ligatures, `smcp` = Small Capitals, etc

I suspect some of the substitutions will be mutually exclusive in most fonts, eg: I doubt fonts have a small-caps variant of tabular figures for example.

With regards to `lvfntman.cpp`, I was just looking at that, and was wondering if one place to start would be to create a single `hb_feature_t` array for each font that combines both `_hb_features` and `_hb_light_features`, with maybe a set of DEFINES for the index of each feature. You could then `features[INDEX].value = 1/0` to enable/disable features as necessary. and add a function or two to enable specific classes (eg, full, light).

Not sure if the above idea would work with your caching, but if it does, it would allow for more flexible feature enabling.

Just a few ideas, Hopefully it might give you some ideas.

**Comment by roger64** [comment]
Hi

1. - Here is a link to a 4-letters tags table if it can be of any help.
https://wiki.documentfoundation.org/Smart_font_optional_features_for_Graphite_and_OpenType_fonts

As you can see, it's an extensive list even if Graphite ones can be dropped.

I think there could be an open discussion on which features to support. My bet, as a common user would be about five. But a mathematician/scientist may wish more...

2. - This other link is also quite useful to check which OT features are available on any font.
https://wakamaifondue.com/

**Comment by poire-z** [comment]
Also https://fontdrop.info/
(And we have enough links about OT features :) what we don't have is that font-variant named values to these tags mapping.)

@shermp : thanks for the input and ideas. Can't comment much for now, not actively thinking about it, busy with other things :)

**Comment by shermp** [comment]
Ok, since I'm thinking about it, I've come up with what I believe should be a mostly complete mapping of CSS to OT names. Only thing missing is stacked fractions, which I can't seem to find/figure out.

| CSS font-variant	| OT Feature tag 	|
|-----------------------|-----------------------|
| contextual 		| 'clig'		|
| historical-ligatures	| 'hlig'		|
| discretionary-lig.	| 'dlig'		|
| common-ligatures	| 'liga' (I think)	|
| lining-nums		| 'lnum'		|
| oldstyle-nums		| 'onum'		|
| proportional-nums	| 'pnum'		|
| tabular-nums		| 'tnum'		|
| diagonal-fractions	| 'frac', ''afrc', 'numr', 'dnom' |
| jis*nn*		| 'jp*nn*		|
| simplified 		| 'smpl'		|
| traditional		| 'trad'		|
| full-width		| 'fwid'		|
| proportional-width	| 'pwid'		|

Hopefully this will help for when or if you get around to thinking of this in the future.

I know you've been very busy on the RTL stuff, so no rush :)

**Comment by poire-z** [comment]
Thanks!
That's really all there is? (~~But you're missing the most popular one: `small-caps`~~ that you just added :)
I didn't see it last times I looked at the specs, and thought each of them can only be enabled - and there's nothing to disable them - which would make handling easier. But I just saw 4 of them can have a `no-` prefix: `no-common-ligatures` `no-discretionary-ligatures` `no-historical-ligatures` `no-contextual`.

**Comment by shermp** [comment]
Yeah, which is why I suggested having a single flexible `hb_feature_t` array would make things a lot simpler if you can make it work.

**Comment by shermp** [comment]
Note that a lot of fonts will not include many of these features, although hopefully harfbuzz will ignore any non-existent features.

**Comment by poire-z** [comment]
Yeah, I guess we should have a pointer to a hb_feature_t instead of a fixed array per font, and may be these full and light hb_feature_t could be made global and shared by all fonts.
When some font-variant stuff needs to adapt it, make a new one dedicated to that font.
But (after thinking about your ealier suggestion) I'm not fond of hardcoding them all - to duplicate what HB would initially use internally, as we'd need to keep that in-sync in case HB changes their defaults.

How come you became interested in that feature (which feels niche, to me)?

**Comment by shermp** [comment]
I really only care about two (maybe three), small-caps and oldstyle figures. Small caps is fairly obvious, and has been a major bugbear of mine about Adobes RMSDK not supporting them.

Old style numbers are surprisingly pleasing to read long-form content with. With my current font, I ended up using FontForge to manually replace the standard numbers with their oldstyle counterparts, but that's a PITA to do.

Fractions might be nice too, but not too worried about them.

As a disclaimer, I don't actually use koreader myself (I dabble occasionally, but I don't like using directory structures for browsing. Databases were invented for a reason!). However, I am very interested in helping improve typography in open source projects in any way I can, because it's all too common to see attitudes of "the basics work, it's good enough" in the realm of text layout/rendering.

As it is, the state of CRengine here is getting to the point of me considering switching away from RMSDK in spite of koreader. I commend anyone who can get me switching away from RMSDK!

**Comment by roger64** [comment]
I can only but concur to what Sherman said. 
Some time ago, I used Prince PDF to convert my ePub, once I had customized them to my liking. I realized a few months ago, with the nice support of footnotes and many other features, that ePub reading with Koreader had become more comfortable than PDF, the only exception I can cite being OT features.  As far as these ones are concerned, Prince support them today in a particular way:  you need to group all of them in the same CSS line, this way...

`font-variant: prince-opentype(ccmp, liga, onum, hlig);

Regarding your question about syntax, the best info I can think of is this very nice Adobe article:
https://helpx.adobe.com/fonts/using/open-type-syntax.html#onum

I think this article covers the major points.

Again, thanks to the Koreader team for the decisive progress that has been made these last months.

`

**Comment by roger64** [comment]
If this can help, this is how LibreOffice implemented a GUI for OT support (from version 6.2 on)

![fonctionnalités](https://user-images.githubusercontent.com/1344670/76483173-46bad780-6451-11ea-8ec6-5d7cb20b7c87.png)
 
![optionOT](https://user-images.githubusercontent.com/1344670/76483211-605c1f00-6451-11ea-8583-5ce8045b9437.png)

**Comment by poire-z** [comment]
Well, ok, given the confinement circumstances, I guess I'm now actively thinking about this request :)

**Comment by poire-z** [comment]
OK, I guess we'll be able to fit all the font-variant named values mentionned in https://drafts.csswg.org/css-fonts-3/#propdef-font-variant-ligatures (and follow ups font-variant-*), except a few ones:
```c
// OpenType features to request
// Max 31 bits. We need "signed int features" to have -1 for non-instantiated fonts)
LFNT_OT_FEATURES_NORMAL 0x00000000

// LFNT_OT_FEATURES_P_LIGA 0x000000XX // +liga +clig enabled by default  (font-variant-ligatures: common-ligatures)
// LFNT_OT_FEATURES_P_CALT 0x000000XX // +calt       enabled by default  (font-variant-ligatures: contextual)
LFNT_OT_FEATURES_M_LIGA 0x00000001 // -liga -clig   (font-variant-ligatures: no-common-ligatures)
LFNT_OT_FEATURES_M_CALT 0x00000002 // -calt         (font-variant-ligatures: no-contextual)
LFNT_OT_FEATURES_P_DLIG 0x00000004 // +dlig         (font-variant-ligatures: discretionary-ligatures)
LFNT_OT_FEATURES_M_DLIG 0x00000008 // -dlig         (font-variant-ligatures: no-discretionary-ligatures)
LFNT_OT_FEATURES_P_HLIG 0x00000010 // +hlig         (font-variant-ligatures: historical-ligatures)
LFNT_OT_FEATURES_M_HLIG 0x00000020 // -hlig         (font-variant-ligatures: no-historical-ligatures)

LFNT_OT_FEATURES_P_HIST 0x00000040 // +hist         (font-variant-alternates: historical-forms)
LFNT_OT_FEATURES_P_RUBY 0x00000080 // +ruby         (font-variant-east-asian: ruby)

LFNT_OT_FEATURES_P_SMCP 0x00000100 // +smcp         (font-variant-caps: small-caps)
LFNT_OT_FEATURES_P_C2SC 0x00000200 // +c2sc +smcp   (font-variant-caps: all-small-caps)
LFNT_OT_FEATURES_P_PCAP 0x00000400 // +pcap         (font-variant-caps: petite-caps)
LFNT_OT_FEATURES_P_C2PC 0x00000800 // +c2pc +pcap   (font-variant-caps: all-petite-caps)
LFNT_OT_FEATURES_P_UNIC 0x00001000 // +unic         (font-variant-caps: unicase)
LFNT_OT_FEATURES_P_TITL 0x00002000 // +titl         (font-variant-caps: titling-caps)
LFNT_OT_FEATURES_P_SUPS 0x00004000 // +sups         (font-variant-position: super)
LFNT_OT_FEATURES_P_SUBS 0x00008000 // +subs         (font-variant-position: sub)

LFNT_OT_FEATURES_P_LNUM 0x00010000 // +lnum         (font-variant-numeric: lining-nums)
LFNT_OT_FEATURES_P_ONUM 0x00020000 // +onum         (font-variant-numeric: oldstyle-nums)
LFNT_OT_FEATURES_P_PNUM 0x00040000 // +pnum         (font-variant-numeric: proportional-nums)
LFNT_OT_FEATURES_P_TNUM 0x00080000 // +tnum         (font-variant-numeric: tabular-nums)
LFNT_OT_FEATURES_P_ZERO 0x00100000 // +zero         (font-variant-numeric: slashed-zero)
LFNT_OT_FEATURES_P_ORDN 0x00200000 // +ordn         (font-variant-numeric: ordinal)
LFNT_OT_FEATURES_P_FRAC 0x00400000 // +frac         (font-variant-numeric: diagonal-fractions)
LFNT_OT_FEATURES_P_AFRC 0x00800000 // +afrc         (font-variant-numeric: stacked-fractions)

LFNT_OT_FEATURES_P_SMPL 0x01000000 // +smpl         (font-variant-east-asian: simplified)
LFNT_OT_FEATURES_P_TRAD 0x02000000 // +trad         (font-variant-east-asian: traditional)
LFNT_OT_FEATURES_P_FWID 0x04000000 // +fwid         (font-variant-east-asian: full-width)
LFNT_OT_FEATURES_P_PWID 0x08000000 // +pwid         (font-variant-east-asian: proportional-width)
LFNT_OT_FEATURES_P_JP78 0x10000000 // +jp78         (font-variant-east-asian: jis78)
LFNT_OT_FEATURES_P_JP83 0x20000000 // +jp83         (font-variant-east-asian: jis83)
LFNT_OT_FEATURES_P_JP04 0x40000000 // +jp04         (font-variant-east-asian: jis04)
// No more room for (let's hope it's really the default in fonts):
// #define LFNT_OT_FEATURES_P_JP90 0x80000000 // +jp90      (font-variant-east-asian: jis90)
```
So, one japanese variant is left out (unless you think another one deserves that more), and these ones that take params and that I'm happy I don't have to think about :)
```
//       - CSS font-variant-alternates:
//       stylistic()                (needs arg, not supported)
//       styleset()                             "
//       character-variant()                    "
//       swash()                                "
//       ornaments()                            "
//       annotation()                           "
```

Still need to think about inheritance:
- full bitmap reset/overriden when a new one is applied from a style
- or inherited and new are OR'ed

Specs says a font-variant reset an upper font-variant, but as we merge all font-variant, font-variant-ligatures, font-variant-caps (and may be later font-feature-settings) into a single bitmap, I'm inclined to go with OR'ing, so we can use style-tweaks like `body { font-variant: oldstyle-nums; }` without that being reset by a lower `H1 { font-variant: small-caps; }`.
(And then, what with `!important` :)

Some samples with various fonts (results really depend on the font, look how/what it decides to handle with diagonal-fraction and ordinal):

Baskervaldx
<kbd>![image](https://user-images.githubusercontent.com/24273478/76886529-38b9ec00-6881-11ea-9108-091d3d48a9ee.png)</kbd>

Fedra Serif
<kbd>![image](https://user-images.githubusercontent.com/24273478/76886619-5dae5f00-6881-11ea-9366-f59635b252a4.png)</kbd>

Free Sans (the only one I have, with Free Serif, with stacked fraction, but not much else)
<kbd>![image](https://user-images.githubusercontent.com/24273478/76886723-7a4a9700-6881-11ea-8e12-9943b6196f66.png)</kbd>

Libertinus Serif
<kbd>![image](https://user-images.githubusercontent.com/24273478/76886893-b7168e00-6881-11ea-9e81-234a6eec5f73.png)</kbd>

Myriad Pro
<kbd>![image](https://user-images.githubusercontent.com/24273478/76886997-dd3c2e00-6881-11ea-88b2-66fdeb9ac998.png)</kbd>

PT Centro
<kbd>![image](https://user-images.githubusercontent.com/24273478/76887069-fba22980-6881-11ea-83a6-3eb96e078eb8.png)</kbd>

Times New Roman
<kbd>![image](https://user-images.githubusercontent.com/24273478/76887139-16749e00-6882-11ea-8dce-26ea6cafb703.png)</kbd>

PS: @RMSDK: get ready to pack your bags!

**Comment by shermp** [comment]
Yes, I'd be inclined to OR them because their could be multiple in play.

**Comment by poire-z** [comment]
OK, hopefully implemented in https://github.com/koreader/crengine/pull/334 - but not really much tested (I checked the values I could test with the fonts I have and latin scripts I read :) inheritance, OR'ing... seem to work as I would expect on a few tests, but it's to be validated on real books).

**Comment by emacsomancer** [comment]
I read through this thread a few times - the emphasis here seems to be on supporting epub books that use Open Type features. Which makes sense. 

But since KOReader allows us to set our own fonts, what if I have a book and a font that has Open Type features. Of course, KOReader has styletweaks, so we can also arbitrary set our own features. 

So I just defined the following styletweak:

```
body {
  font-kerning: normal;
  font-variant: oldstyle-nums discretionary-ligatures historical-ligatures historical-forms;
  font-feature-settings: "kern", "dlig", "liga", "hlig", "hist", "onum";
}
```

Moved it into `koreader/styletweaks`, enabled it for the book, then used one of the [IM Fell fonts](https://iginomarini.com/fell/), and here's a page from a regular .epub of *The Compleat Angler* (which doesn't itself call any Open Type features):

![155286879_10165196537955045_3826654271262119831_o](https://user-images.githubusercontent.com/9464083/109738839-d509fb00-7b85-11eb-9d63-99dcb7320eff.jpg)

Just in case it's useful (though there may be better ways of doing this), seemed worth making a brief note of here.

**Comment by roger64** [comment]
Thanks for the tip.

I think the font-feature settings are not - yet - implemented on Koreader.

How do you add a personal style tweak ? I see it's in /koreader/styletweaks but it's a folder. Do you insert a text file, a lua file?

**Comment by emacsomancer** [comment]
@roger64 - I don't see anywhere in KOReader menu to interact with font features, but the styletweaks work current (as above).

So I just made a file called `17th.css` with the content I showed above, i.e.:
```
body {
  font-kerning: normal;
  font-variant: oldstyle-nums discretionary-ligatures historical-ligatures historical-forms;
  font-feature-settings: "kern", "dlig", "liga", "hlig", "hist", "onum";
}
```

(but you might, of course, choose a different set of features; and I indeed also made a different .css file without historical-forms enabled [so then no long ſ in that case].)

And then this goes in the `styletweaks` folder.

Then, in KOReader itself, chose [Document](2nd icon from left)>'Style tweaks'>'User style tweaks'>[click on name, e.g. =17th.css= ].

**Comment by roger64** [comment]
@emacomancer
A css file... I should have guessed it. Thanks!

**Comment by poire-z** [comment]
Regarding https://www.mobileread.com/forums/showthread.php?t=337924

> It seems to me a KOReader bug: the CSS states to use Simoncini with small caps or fallback to Serif with small caps if Simoncini with small caps is not present. KOReader uses Simoncini without small caps and does not fallback to Serif with small caps. 

We did not have support for smallcaps. Turns out having support for OT features would give support for smallcaps if the font has OT support for them. So ok, we could have support for that this way.
Althought we can't really know if the font has support or not for smcp (dunno if we can query that via Hatfbuzz).
I think other renderers, in such a case, fallback not to the 2nd font (which could look ugly), but to using fake/simulated smallcaps with the first font (which might look better) (with something equivalent to `{ font-size: 0.9em; text-transform: uppercase}` ?)
So yes, probably a bug not ensuring them, but not really bothered doing all the gymnastics trying to fix this case.

**Comment by NiLuJe** [comment]
Yeah, that was what my "does it even make sense" comment was alluding to: I'm also of the mind that not randomly choosing another font is a *better* approach ;). Actually faking smallcaps would be the cherry on top, but, yeah, would require being able to query (which [harbuzz can probably do](https://harfbuzz.github.io/harfbuzz-hb-ot-layout.html)).

**Comment by poire-z** [comment]
https://www.mobileread.com/forums/showthread.php?p=4101536#post4101536 @Jellby 

> There is another question:
Can KOReader create "synthetic" smallcaps, italics and bold? Some renderers do, and can artificially slant or embolden a font, or automatically display "fake" smallcaps. Synthetic italic and bold are difficult (especially italic, which can only be slanted), but synthetic fake smallcaps are easy: just convert lowercase to uppercase and display smaller.
The following question is:
If a renderer can generate synthetic smallcaps, and a given requested font is available but does not have true smallcaps, should the renderer create the fake smallcaps, or should it fall back to the next font in the list? The same for italic and bold, of course. 

crengine does fake bold and fake italic for fonts that don't come with any such variants - and quite well in my opinion. But all that is "easy", as it's just a unique font, from which we get glyphs, and apply some transformation to the glyphs: the advance and position of the original glyphs are kept (so, our fake bold looks bolder, but is not wider than the same text with the original normal font).
https://github.com/koreader/crengine/pull/309 (there's been a few corrections/improvements to fake bold or italic in some later PRs I think)

For small caps, it's a lot more complicated: we can do text transform, but we would need to use multiple font instances - and there it gets really complicated...
In crengine, we have a HTML node, and the text node it contains. We get THE font to use from THE HTML node, and we ask THIS font to measure and draw that text content (possibly text-transform'ed, to uppercase or lowercase, but this ends up being given to THE font as THE text to measure and draw). We never have to do any other way.
Doing "fake small cap" would need to cut thie text content into fragments (the lowercase parts, the uppercase parts) and draw each fragment with a different font. This has implications in many places, and would need to add much kludge to handle this specific case.
It might be easy for other renderers - but it is not for crengine.

**Comment by shermp** [comment]
What about doing it at the xhtml level? Basically, add spans around the appropriate text fragments, set appropriate css (text-transform, and font size), then pass that to the renderer?

**Comment by poire-z** [comment]
Because the DOM is sacred :) We may add/insert "boxing" internal elements at load time (for more serious purposes than smallcaps), but then, they stick and are sacred too.
After that, we may change styles, and most will apply on this DOM. That means that if I want `* { font-variant: smallcaps }`,  I would need for this to work to "box" EVERY segment of lowercase letters, then every bit of uppercase letters, then...
And when I apply `* { font-family: inherit !important }` so my main OT font with OT smallcap is used, it will see the many uneeded fragments, and shape each segment individually, making everything slow and a mess :)

Eventually, if we'd have to, it probably be best done all in the font code:  THE font instance could know if it supports +smcp, and if not, uppercase/lowercase each glyph appropriately, and may be scale the glyph to make it smaller. But then, Harfbuzz do its own thing with the font...

Anyway, we might not do it good by default - but a user who wants them can help himself (with this PR) by installing a complete OT aware Simoncini font - and just disable Embedded Fonts so his local font is used instead of the EPUB one.

**Comment by shermp** [comment]
Fair enough. Indeed, probably too much work, especially when CRe actually supports true small-caps.

**Comment by Jellby** [comment]
> For small caps, it's a lot more complicated: we can do text transform, but we would need to use multiple font instances - and there it gets really complicated...

What's easy for a human is not necessarily easy for a computer ;) I meant that fake bold/italic involve changing the glyph shapes, while fake smallcaps would involve (naïvely) just replacing lowercase characters with scaled-down versions of their uppercase counterparts. Of course, things get complicated when ligatures, kerning, and other OT features start popping up (e.g., the "av" kerning should be replaced by the scaled-down "AV" kerning...).

I wonder if CRe could some other modification to the font to indicate that it should be small-caps, even if it doesn't look

**Comment by poire-z** [comment]
Been testing some new fonts, and it happens some of then didn't have OT smallcaps support, while I was reading a book with lots of `xvii`<sup>e</sup> (siècle/century) that were a bit (but not that much) distracting.
I again had a look at how we could do it, but no clean solution jumped at me :/
Just dropping some URLs I met while looking again (for when I feel like looking again...):

`https://github.com/latex3/luaotfload/issues/116#issuecomment-551091045` _hard to do right, nothing convincing_
https://github.com/mozilla/gecko-dev/blob/master/gfx/thebes/gfxFont.cpp#L3260 Firefox fake small caps
https://github.com/mozilla/gecko-dev/blob/master/gfx/thebes/gfxFontEntry.cpp#L862 Firefox detection of OT smallcap support, using bits from https://harfbuzz.github.io/harfbuzz-hb-ot-layout.html

**Comment by Frenzie** [comment]
I didn't realize Fx had some custom thing instead of just depending on FreeType or something. (Or are those synthetic* things from FT?)

**Comment by poire-z** [comment]
From what I could grasp; no. They convert lowercase sequences to uppercase, and run that thru a 0.8em font. They don't seem to do any glyph thickness correction.
This conversion/font switch seems like not much, but it's complicated for us because we have 3 rendering methods (freetype, harfbuzz light, harfbuzz full) which do things differently, and we switch internally in these runs processing to fallback fonts that may have different support for OT smallcaps.
(Might be easier for the non-harfbuzz methods, but I guess it's not worth the work as most people do use harfbuzz/best.)

**Comment by Frenzie** [comment]
> They convert lowercase sequences to uppercase, and run that thru a 0.8em font. They don't seem to do any glyph thickness correction.

Interesting. Well, I suppose thickness correction doesn't matter much because the dimensions will look slightly off regardless.

**Comment by Jellby** [comment]
I wonder if it would be better to scale by `1ex`/`1em`, instead of a fixed `0.8` (in case it's ever implemented).

**Comment by Frenzie** [comment]
Yes, that should offer better emulation.

**Comment by poire-z** [comment]
Had a try at what I thought could be an easy solution - like the wrap into spans solution, but actually at some inner level when we add text segments to give to the text formatting code.
```diff
--- a/crengine/src/lvrend.cpp
+++ b/crengine/src/lvrend.cpp
@@ -4048,2 +4048,32 @@ void renderFinalBlock( ldomNode * enode, LFormattedText * txform, RenderRectAcce
             if ( txt.length()>0 ) {
+                if ( style->font_features.value & LFNT_OT_FEATURES_P_SMCP ) {
+                    LVFontRef altfnt = fontMan->GetFont( font->getSize() * 0.8,
+                        font->getWeight(), font->getItalic(), font->getFontFamily(), font->getTypeFace(),
+                        font->getFeatures(), enode->getDocument()->getFontContextDocIndex(), true);
+                    lString32 utxt = txt;
+                    utxt.uppercase();
+                    int segment_start = 0;
+                    bool segment_differ = txt[0] != utxt[0];
+                    bool emit_segment;
+                    for ( int i=1; i<=txt.length(); i++ ) {
+                        if (i == txt.length()) {
+                            emit_segment = true;
+                        }
+                        else {
+                            bool differ = txt[i] != utxt[i];
+                            emit_segment = differ != segment_differ;
+                        }
+                        if ( emit_segment ) {
+                            lChar32 * t = (lChar32*)(utxt.c_str()+segment_start);
+                            int tlen = i - segment_start;
+                            LVFontRef fnt = segment_differ ? altfnt : font;
+                            txform->AddSourceLine( t, tlen, cl, bgcl, fnt.get(), lang_cfg, baseflags | tflags,
+                                line_h, valign_dy, indent, enode, 0, letter_spacing );
+                            baseflags &= ~LTEXT_FLAG_NEWLINE & ~LTEXT_SRC_IS_CLEAR_BOTH; // clear newline flag
+                            segment_start = i;
+                            segment_differ = !segment_differ;
+                        }
+                    }
+                }
+                else {
                 txform->AddSourceLine( txt.c_str(), txt.length(), cl, bgcl, font.get(), lang_cfg, baseflags | tflags,
@@ -4055,2 +4085,3 @@ void renderFinalBlock( ldomNode * enode, LFormattedText * txform, RenderRectAcce
                 //          lang_cfg, baseflags|tflags|LTEXT_FLAG_OWNTEXT, line_h, valign_dy, 0, enode );
+                }
             }
```

This is what we currently get with Harfbuzz using OT fonts smallcap:
<kbd>![image](https://user-images.githubusercontent.com/24273478/112756626-983ef180-8fd5-11eb-8dd3-a923933181cb.png)</kbd>

This is what we get currently when not using Harfbuzz (font kerning "fast"):
<kbd>![image](https://user-images.githubusercontent.com/24273478/112756656-bad10a80-8fd5-11eb-9a3d-ebbe29a612e2.png)</kbd>

This is what we could get instead with the above patch:
<kbd>![image](https://user-images.githubusercontent.com/24273478/112756675-df2ce700-8fd5-11eb-8a1c-a09ad2ddd14c.png)</kbd>

Unfortunately, text highlighting/selection does not work in there because this part of the code expects a text node to be a single segment... and it's already a quite tedious part of code I'd rather not tweak just for that.
The bottom line is that every idea needs some more or less heavy kludge elsewhere to have the whole thing working :/
